### PR TITLE
[GH-#370] Print of a patricia merkle trie

### LINKED ISSUE #370: pretty print function for patricia merkle tree
@cytadela8 recommended to provide a way to print patricia merkle trees content nicely in the shell, for debugging and logging purposes

### PR DESCRIPTION
**Pull request by meivantodorov**
Providing API for printing a patricia merkle trie in the shell as struct or as key value pair

Closes #370 

Merge after #366

### CODE REVIEW

**Comment by thepiwo** [comment]
how can I call this? `Aeutil.PatriciaMerkleTree.print_trie(Chain.top_block_chain_state.accounts)` doesn't work for me

**Comment by meivantodorov** [comment]
> can you add PatriciaMerkleTree to the iex aliases?

Thanks, I will.

> how can I call this? Aeutil.PatriciaMerkleTree.print_trie(Chain.top_block_chain_state.accounts) doesn't work for me.

You need #366 to be merged first. Until then You can make simple trie and test it until then.Like:
```
 t = Aeutil.PatriciaMerkleTree.new(:account) |> Aeutil.PatriciaMerkleTree.enter("k1", "v1") |> Aeutil.PatriciaMerkleTree.enter("k2", "v2")
 Aeutil.PatriciaMerkleTree.print_trie t
```

**Comment by thepiwo** [comment]
How will I do it for the accounts tree in chain?

**Comment by thepiwo** [comment]
@meivantodorov How will I do it for the accounts tree in chain?

**Comment by thepiwo** [comment]
@meivantodorov could you have another look at this?

**Comment by cytadela8** [comment]
I did the supplied example and got following results:

```
iex(7)> PatriciaMerkleTree.print_trie t                                                                                                            
~~~Trie~~~
Node: ext (prefix: [3, 1, 3, 1, 3])
  Node: branch (value: "")
    [0] Node: <empty>
    [1] Node: leaf ([]="v1")
    [2] Node: leaf ([]="v2")
    [3] Node: <empty>
    [4] Node: <empty>
    [5] Node: <empty>
    [6] Node: <empty>
    [7] Node: <empty>
    [8] Node: <empty>
    [9] Node: <empty>
    [10] Node: <empty>
    [11] Node: <empty>
    [12] Node: <empty>
    [13] Node: <empty>
    [14] Node: <empty>
    [15] Node: <empty>
~~~/Trie/~~~
```

So this is a step in good direction, but it's still not what I intended when suggesting this feature to thepiwo.

The current problem is that the state off all accounts is stored in merkle tree and that's not easily readable. Sometimes it's useful to be able to print the states off all accounts. In future when we change to merkle trees for all parts of chainstate this problem will also apply to oracles, namings, channels, etc.

To be able to easily examine contents of a merkle tree I would suggest we should create a function that takes a merkle tree as stored in chainstate and prints all pairs: leaf path - value. Optionally we could also create a function that takes a merkle tree and returns all keys (leaf paths).

_note: of course currently only accounts are stored in a merkle tree, but in future other parts of chainstate will also be_

**Comment by meivantodorov** [comment]
@cytadela8 When the all chainstates are changed from merkle trie to patricia merkle trie You can then print what You want. This is not merkle trie print functionality. There is an issue for getting all keys of a given patricia merkle trie.

**Comment by thepiwo** [comment]
the task is supposed to "provide a way to print patricia merkle trees content nicely"

**Comment by meivantodorov** [comment]
> the task is supposed to "provide a way to print patricia merkle trees content nicely"

"Nicely" keeping the structure of the trie, or printing the keys and the values, or different way of "nicely" ?

**Comment by cytadela8** [comment]
I would consider nicely as printing keys and values pairs. That's what I meant when I proposed this task.

**Comment by meivantodorov** [comment]
@cytadela8 
> I would consider nicely as printing keys and values pairs.

Thank You, I will do it so as it will be more helpful.

**Comment by thepiwo** [comment]
Sorry for the misunderstanding, with nicely we meant easily human readable, so the content is understandable, not the structure

**Comment by thepiwo** [comment]
@meivantodorov can you give an update on the status?

File: apps/aeutil/lib/patricia_merkle_tree.ex [hunk 1]
```diff
@@ -86,12 +87,54 @@ defmodule Aeutil.PatriciaMerkleTree do
...
   def delete(trie, key), do: Trie.delete(trie, key)
+
+  @doc """
+  Providing pretty print of a given trie in the shell.
+  Depending on the atom it can print structure or key value pairs
+
+  ## Examples
+
+  If we want to print as pair
+      iex> Aeutil.PatriciaMerkleTree.new(:test_trie) |> Aeutil.PatriciaMerkleTree.enter("111", "val1") |> Aeutil.PatriciaMerkleTree.enter("112", "val2") |> Aeutil.PatriciaMerkleTree.print_trie(:as_pair)
```
Comment (thepiwo) [inline apps/aeutil/lib/patricia_merkle_tree.ex]: I think one example to print one chain state tree would be sufficient

Comment (meivantodorov) [reply apps/aeutil/lib/patricia_merkle_tree.ex]: > I think one example to print one chain state tree would be sufficient

After #366 is merged, I can provide such an example.

**Comment by thepiwo** [comment]
will close this util unblocked and resolved

**Pull request closed without merging**